### PR TITLE
feat(clickhouse)!: Upgrade to `github.com/cloudquery/plugin-sdk/v2`

### LINKED ISSUE #9847: feat: Migrate `plugin/destination/clickhouse` to `github.com/cloudquery/plugin-sdk/v2`
### Which problem is this feature request solving?

Part of #9846

### Describe the solution you'd like

Upgrade the code performing the major bump.
PR description should include the following part:

```
BEGIN_COMMIT_OVERRIDE
feat: Update to use [Apache Arrow](https://arrow.apache.org/) type system

BREAKING-CHANGE: This release introduces an internal change to our type system to use [Apache Arrow](https://arrow.apache.org/). This should not have any visible breaking changes, however due to the size of the change we are introducing it under a major version bump to communicate that it might have some bugs that we weren't able to catch during our internal tests. If you encounter an issue during the upgrade, please submit a [bug report](https://github.com/cloudquery/cloudquery/issues/new/choose).

END_COMMIT_OVERRIDE
```

### Pull request (optional)

- [X] I can submit a pull request

### PR DESCRIPTION
**Pull request by candiduslynx**
Closes #9847 

BEGIN_COMMIT_OVERRIDE
feat: Update to use [Apache Arrow](https://arrow.apache.org/) type system (#10284)

BREAKING-CHANGE: This release introduces an internal change to our type system to use [Apache Arrow](https://arrow.apache.org/). This should not have any visible breaking changes, however due to the size of the change we are introducing it under a major version bump to communicate that it might have some bugs that we weren't able to catch during our internal tests. If you encounter an issue during the upgrade, please submit a [bug report](https://github.com/cloudquery/cloudquery/issues/new/choose).

feat: Allow nullable columns in primary keys (#10284)
BREAKING-CHANGE: This change enables [`allow_nullable_key`](https://clickhouse.com/docs/en/operations/settings/settings#allow-nullable-key) for tables (#10284).

END_COMMIT_OVERRIDE

### CODE REVIEW

**Review by yevgenypats — CHANGES_REQUESTED** [review]
Just marking so I've a chance to review before it is merged.

**Comment by candiduslynx** [comment]
@yevgenypats @erezrokah 
As some code had to be copied/based on ClickHouse Go SDK, I think we might need to include this info per license requirements:
https://github.com/ClickHouse/clickhouse-go/blob/main/LICENSE

File: plugins/destination/clickhouse/typeconv/ch/values/list.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		if elem != nil {
+			val.Elem().Set(reflect.ValueOf(elem))
+		}
+		res.Index(i).Set(val)
+	}
+
+	return res.Interface(), nil
+}
+
+type listWrapper struct {
```
Comment (candiduslynx) [inline plugins/destination/clickhouse/typeconv/ch/values/list.go:52]: I really think this should be implemented in the Arrow Go code (`*array.FixedSizeList` should conform to the `array.ListLike` interface)

Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/values/list.go:52]: Proposed in https://github.com/apache/arrow/issues/35325, PR: https://github.com/apache/arrow/pull/35327

File: plugins/destination/clickhouse/typeconv/arrow/types/field.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	"fmt"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2/lib/column"
+	"github.com/apache/arrow/go/v12/arrow"
+	"github.com/cloudquery/cloudquery/plugins/destination/clickhouse/util"
+	"github.com/cloudquery/plugin-sdk/v2/types"
+)
+
+func fieldFromColumn(col column.Interface) (*arrow.Field, error) {
```
Comment (yevgenypats) [inline plugins/destination/clickhouse/typeconv/arrow/types/field.go:13]: why not just use the string representation?

File: plugins/destination/clickhouse/typeconv/arrow/types/field.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	name := util.UnquoteID(col.Name())
+	switch col := col.(type) {
+	case *column.Bool:
+		return &arrow.Field{Name: name, Type: new(arrow.BooleanType)}, nil
```
Comment (yevgenypats) [inline plugins/destination/clickhouse/typeconv/arrow/types/field.go:17]: there is no need to create those types. take a look here - https://github.com/cloudquery/cloudquery/blob/main/plugins/destination/postgresql/client/types_pg.go#L29

File: plugins/destination/clickhouse/typeconv/ch/values/time.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/apache/arrow/go/v12/arrow"
+	"github.com/apache/arrow/go/v12/arrow/array"
+)
+
+type date interface {
+	ToTime() time.Time
+}
+
+func dateValue[A date](arr primitive[A]) []*time.Time {
```
Comment (yevgenypats) [inline plugins/destination/clickhouse/typeconv/ch/values/time.go:14]: what's the point of generics here? can't it be just an interface ?

File: plugins/destination/clickhouse/typeconv/ch/values/decimal.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+	"github.com/apache/arrow/go/v12/arrow"
+	"github.com/shopspring/decimal"
+)
+
+type decimalToBigInt interface {
+	BigInt() *big.Int
+}
+
+func decimalValue[A decimalToBigInt](arr primitive[A]) []*decimal.Decimal {
```
Comment (yevgenypats) [inline plugins/destination/clickhouse/typeconv/ch/values/decimal.go:14]: Same. what the point of generics here? can it be just an interface?

File: plugins/destination/clickhouse/typeconv/ch/types/data_type.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
+package types
+
+import (
+	"strconv"
+
+	"github.com/apache/arrow/go/v12/arrow"
+)
+
+func dataType(_type arrow.DataType) (string, error) {
+	switch _type.ID() {
```
Comment (yevgenypats) [inline plugins/destination/clickhouse/typeconv/ch/types/data_type.go:10]: seems there is a mix between usage of `ID` and type casting. can we use just one? similar to https://github.com/cloudquery/cloudquery/blob/main/plugins/destination/postgresql/client/types_pg.go#L29 ?

**Review by yevgenypats — CHANGES_REQUESTED** [review]
lot's of code :) few initial comments - will take a break and then give it another fresh look tomorrow.

I did an initial research and seems this should be our best bet in clickhouse indeed in terms of transformation (transforming to json or to arrow will be too slow and we might loose some control that we need)

File: plugins/destination/clickhouse/typeconv/arrow/types/field.go (hunk 2 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/arrow/types/field.go:13]: This allows us not to parse a lot ourselves, and instead we just handle the SDK types.
Some (e.g., `Tuple`, precision & time zone for `DateTime64`, etc.) parts are still required to be parsed, but mostly we just can lean on the SDK to unwrap the nested types for us.

File: plugins/destination/clickhouse/typeconv/arrow/types/field.go (hunk 3 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/arrow/types/field.go:17]: Could you elaborate?

File: plugins/destination/clickhouse/typeconv/ch/values/time.go (hunk 4 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/values/time.go:14]: We're inferring the interface of the value returned by `Value(int)`. In this case, it's either `arrow.Date32` & `arrow.Date64`.
If we were to use interfaces there would've been 2 separate functions (we might as well just do 2 separate functions for each array type).

File: plugins/destination/clickhouse/typeconv/ch/values/decimal.go (hunk 5 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/values/decimal.go:14]: Same as Date32/64 but for Decimal128/256

File: plugins/destination/clickhouse/typeconv/ch/types/data_type.go (hunk 6 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/types/data_type.go:10]: Good point, I'll look into it

File: plugins/destination/clickhouse/typeconv/ch/values/list.go (hunk 1 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/values/list.go]: merged, updated the code in 6cda3a3

File: plugins/destination/clickhouse/typeconv/ch/types/data_type.go (hunk 6 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/ch/types/data_type.go]: Done in fb53fc1

File: plugins/destination/clickhouse/typeconv/arrow/types/field.go (hunk 3 above)
Comment (candiduslynx) [reply plugins/destination/clickhouse/typeconv/arrow/types/field.go:17]: This is the reverse type transformation (from ClickHouse to Apache Arrow) as it's required to perform comparison in the normalized schemas.

**Review by yevgenypats — CHANGES_REQUESTED** [review]
I think let's try go easy on generics. maybe we should disable generics use across all our repos.

I don't see an immediate need for generics in this PR. For example all generics code under `values` can be replaces with one call to `.Values` available on all arrow arrays (or `.Uint16Values` for some reason it is called by different name sometimes but nevertheless it is available there so should save lots of code and test code in this PR which will make it hard to maintain)

**Comment by candiduslynx** [comment]
> all generics code under `values` can be replaces with one call to `.Values` available on all arrow arrays

@yevgenypats are you referring to the `ch/values` package?
If so, the reason for the generics is that we do need concrete type for the values passed to the ClickHouse SDK, and the `Value` method for different array types returns different types altogether (not `any`).

If we take a look at the `arrow/values` and generics usage there, removing them would mean going for the separate functions per array (`Append` takes concrete types once again).

Could you give an example on what do you suggest doing and where?

**Review by yevgenypats — APPROVED** [review]
Going to stamp approve it.  Might go back to that if time allows.

**Pull request merged**